### PR TITLE
treewide: drop LLVM11

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

unpin llvm11 from freshbootstraptools and drop LLVM11. i have a PR in review to  update the bootstrap-tools for darwin which should be merged soonish https://github.com/NixOS/nixpkgs/pull/295557

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by a-n-n-a-l-e-e** [comment]
as per @reckenrode
> Bootstrap tools should probably be the minimally supported version of LLVM to prevent accidentally depending on the bootstrap version of LLVM matching the version in the stdenv.

so that will need to be changed from llvmPackages_16.

**Comment by reckenrode** [comment]
To elaborate on that a bit, I want to avoid assumptions about the bootstrap compiler lest it cause trouble for future LLVM version updates for Darwin. In theory, it only matters if the bootstrap tools are ever updated, and the breakage would have to happen between now and when the switch is made to LLVM 18., so the risk is low. It may not be worth changing the version if it creates more work. (Assuming the SDK issue with libc++ can be addressed.)

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -20173,8 +20158,6 @@ with pkgs;
   swig = swig3;
   swigWithJava = swig;
 
-  c2ffi = callPackage ../development/tools/misc/c2ffi { };
```
Comment (alyssais) [inline pkgs/top-level/all-packages.nix:20176]: Gentoo has a version from 2023 — would it be easy enough to just update this rather than dropping it?

**Comment by elliottslaughter** [comment]
Please be aware that if you remove LLVM 11, you will need to disable Terra on Linux ARM due to this outstanding issue: https://github.com/terralang/terra/issues/597

On every system other than Linux ARM, Terra can go up to LLVM 17 if needed, but the Linux ARM bug keeps that platform on LLVM 11.

**Comment by a-n-n-a-l-e-e** [comment]
> Please be aware that if you remove LLVM 11, you will need to disable Terra on Linux ARM due to this outstanding issue: [terralang/terra#597](https://github.com/terralang/terra/issues/597)
> 
> On every system other than Linux ARM, Terra can go up to LLVM 17 if needed, but the Linux ARM bug keeps that platform on LLVM 11.


<details>
<summary>the build fails when trying to compile with an llvm above 13. eg, with llvm16</summary>
<pre>
CMake Error at src/CMakeLists.txt:229 (target_link_libraries):
  Target "TerraLibraryShared" links to:

    FFI::ffi

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at src/CMakeLists.txt:345 (target_link_libraries):
  Target "TerraExecutable" links to:

    FFI::ffi

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
</pre>
</details>

**Comment by elliottslaughter** [comment]
@a-n-n-a-l-e-e let me know if I should move this discussion to a different thread; until then I'll just reply here.

That looks like a missing dependency. In general, Terra only relies on a minimal LLVM build. (For our builds, I usually turn off all the optional build flags.) It's fine if you built LLVM with optional dependencies but then that means you'll have to add those dependencies to Terra.

I'm not an expert on what this `FFI` is but I've noticed that in many distros, the set of default LLVM dependencies has been creeping up over time (even though LLVM actually does not require them to be enabled). I would consult the LLVM build itself to determine what this `FFI` is and then add it to Terra, and see if that fixes it.

I see I forgot to release the current support for LLVM 17 in Terra, which I'll probably do shortly. Until then, version 1.1.1 supports up to LLVM 16.

**Comment by a-n-n-a-l-e-e** [comment]
thanks @elliottslaughter. PR updated with llvm16. just needed to add libffi and update a path.

**Pull request merged**